src/kernel/sporadic.c

```diff
@@ -227,47 +227,27 @@ static inline void schedule_used(sched_context_t *sc, refill_t new)
     assert(!refill_empty(sc));
 }
 
-void refill_budget_check(ticks_t usage, ticks_t capacity)
+void refill_budget_check(ticks_t usage)
 {
     sched_context_t *sc = NODE_STATE(ksCurSC);
     /* this function should only be called when the sc is out of budget */
-    assert(capacity < MIN_BUDGET || refill_full(sc));
     assert(sc->scPeriod > 0);
     REFILL_SANITY_START(sc);
 
-    if (capacity == 0) {
```

Comment:
I don't know if this change is correct or not and I don't have time to go through it. Can you explain why this can be removed?

Reply:
The initial loop will charge any complete refills as if they had been used which preserves future bandwidth in the presence of an overrun. The second part delays a refill that isn't completely used by the amount of time overrun. Delaying may work for the logic of sporadic server which does not attempt to preserve temporal isolation but only delaying will not preserve bandiwdth as it can still result in a subsequent period window with greater than proportional utilisation. Dealing with a partially consumed refill in an overrun by simply using `refill_split_check` will charge it in the exact same manner as if it had been consumed which should actually preserve future bandwidth in the event of overrun.

Reply:
ok. Can you add this to a comment stating why it's different with reference to the paper? Might be a good idea to run this via Björn